### PR TITLE
CDAP-19023: add support for configuring composite trigger

### PR DESCRIPTION
**Pull request by de-lan**
This PR adds support for configuring composite trigger, more specifically it adds the id information for the respective composite trigger. Originally schedule trigger was default to the PROGRAM_STATUS type when parsing runtime arguments. The sister UI pr is here: https://github.com/cdapio/cdap-ui/pull/414

### CODE REVIEW

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/SmartWorkflow.java
Comment (de-lan) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/SmartWorkflow.java]: This function behaves exactly like `updateTokenWithTriggeringProperties`, except that it iterates through all the triggers and match the id information

File: cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/ArgumentMapping.java
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/ArgumentMapping.java]: don't need to do this, just do `+ pipelineId` like it's done for source and target.

File: cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/ArgumentMapping.java
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/ArgumentMapping.java]: indentation should be 2 spaces, make sure your IDE has the right settings.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/test/java/io/cdap/cdap/datapipeline/DataPipelineTest.java
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/test/java/io/cdap/cdap/datapipeline/DataPipelineTest.java]: indentation

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/test/java/io/cdap/cdap/datapipeline/DataPipelineTest.java
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/test/java/io/cdap/cdap/datapipeline/DataPipelineTest.java]: this doesn't really test that the logic is correct. Should have a test where both pipelines have the same property name in order to make sure the triggering pipeline name is taken into account correctly

File: cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/PluginPropertyMapping.java
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/PluginPropertyMapping.java]: similar comment, don't need the null check

File: cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/PluginPropertyMapping.java
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/PluginPropertyMapping.java]: this should be updated to call the other constructor

```
this(stageName, source, target, null)
```

File: cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/TriggeringPipelineId.java
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/TriggeringPipelineId.java]: indentation

File: cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/TriggeringPipelineId.java
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/TriggeringPipelineId.java]: indentation

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/SmartWorkflow.java
Comment (albertshau) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/SmartWorkflow.java]: should not copy-paste code like this. Refactor things so that this method calls updateTokenWithTriggeringProperties.

While you're at it, I think it would also be a good idea to change updateTokenWithTriggeringProperties so that it is a static method that returns a Map of keys rather than mutating the WorkflowToken within the method. That way you can add unit tests that calls the method directly rather than the current tests that need to deploy and run entire pipelines

File: cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/ArgumentMapping.java
Comment (de-lan) [reply cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/ArgumentMapping.java]: `pipelineId` is a Class instead of String like source and target.  Guess have to do a null check...

Comment (albertshau) [reply cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/ArgumentMapping.java]: if it has a toString() method java will use that, you don't need to do this manually

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/test/java/io/cdap/cdap/datapipeline/DataPipelineTest.java
Comment (de-lan) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/test/java/io/cdap/cdap/datapipeline/DataPipelineTest.java]: will update with a new dedicated test

File: cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/ArgumentMapping.java
Comment (de-lan) [reply cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/ArgumentMapping.java]: gotcha! that's a good point, thanks!

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/SmartWorkflow.java
Comment (de-lan) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/SmartWorkflow.java]: Thanks! updated

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/SmartWorkflow.java
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/SmartWorkflow.java]: style: for readability, we don't do static imports.

Import just the Feature class and refer to it as Feature.PIPELINE_COMPOSITE_TRIGGERS

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/WorkflowScheduleUtils.java
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/WorkflowScheduleUtils.java]: should have a short javadoc for the class

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/WorkflowScheduleUtils.java
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/WorkflowScheduleUtils.java]: nit: let's name it PipelineTriggers so that it's more specific.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/WorkflowScheduleUtils.java
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/WorkflowScheduleUtils.java]: don't pass in a Logger, create a static Logger for this class like in other places. Log messages should be come from a Logger for the class that has the code.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/WorkflowScheduleUtils.java
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/WorkflowScheduleUtils.java]: is this just copy-pasted code? It's a bit hard to follow, so if there are any differences from before can you comment on them?

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/WorkflowScheduleUtils.java
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/WorkflowScheduleUtils.java]: nit: process -> add

This makes it a little more clear that it is adding things to the map.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/WorkflowScheduleUtils.java
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/WorkflowScheduleUtils.java]: this description should have more detail. It should include some example input and what gets added to the mappings.

Somebody should be able to read this and tell what the method does, especially since the code isn't the clearest.

File: cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/ArgumentMapping.java
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/ArgumentMapping.java]: change this to call `this(source, target, null)`

File: cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/TriggeringPipelineId.java
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/TriggeringPipelineId.java]: pipelineName is redundant, can rename this to just `name`.

File: cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/ArgumentMapping.java
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/ArgumentMapping.java]: rename this to just `pipeline`. It's a little more future-proof in case we ever need to add more fields to the TriggeringPipelineId object and it no longer is an id.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/WorkflowScheduleUtils.java
Comment (de-lan) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/WorkflowScheduleUtils.java]: yes, this is partially copied code, but also with some new code checking on the pipeline ID

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/SmartWorkflow.java
Comment (de-lan) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/SmartWorkflow.java]: updated

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/WorkflowScheduleUtils.java
Comment (de-lan) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/WorkflowScheduleUtils.java]: javadoc added

File: cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/ArgumentMapping.java
Comment (de-lan) [reply cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/ArgumentMapping.java]: updaed

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/PipelineTriggersUtils.java
Comment (de-lan) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/PipelineTriggersUtils.java]: Most of the code here is from `SmartWorkflow.java`, with slight difference that now it checks the ID if `compositeTriggerEnabled` is true

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/PipelineTriggersUtils.java
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/PipelineTriggersUtils.java]: nit: we normally don't put 'Utils' in the name, just PipelineTriggers.

**Comment by albertshau** [comment]
for future reference, you need to add the 'build' label in order to tell it to run the build

**Review by albertshau — APPROVED** [review]
lgtm assuming the build passes. Please squash commits before merging.

**Comment by de-lan** [comment]
> for future reference, you need to add the 'build' label in order to tell it to run the build

gotcha, will do in future

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/PipelineTriggersUtils.java
Comment (de-lan) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/PipelineTriggersUtils.java]: Aha, I followed the other files naming convention, will update

**Comment by de-lan** [comment]
Mis-clicked the close button...

**Pull request merged**
